### PR TITLE
Add constbuffer_array_remove_empty_buffers API with comprehensive tes…

### PR DESCRIPTION
**Pull request by dcristoloveanu**
…ting

- Add constbuffer_array_remove_empty_buffers() function to sanitize constbuffer arrays
- Implement SRS requirements with proper reference counting semantics
- Add 126 unit tests covering all edge cases and error conditions
- Add 6 integration tests for real-world scenarios and reference counting validation
- Include reals library support for non-mocked testing scenarios
- All tests pass with VLD memory leak detection enabled

### CODE REVIEW

File: tests/constbuffer_array_int/constbuffer_array_int.c [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+#include "c_util/constbuffer.h"
+#include "c_util/constbuffer_array.h"
+
+BEGIN_TEST_SUITE(constbuffer_array_int)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+    int result = gballoc_hl_init(NULL, NULL);
+    ASSERT_ARE_EQUAL(int, 0, result);
```
Comment (dcristoloveanu) [inline tests/constbuffer_array_int/constbuffer_array_int.c:24]: No need for a temp variable, can just make this a one liner.

File: tests/constbuffer_array_int/constbuffer_array_int.c [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+    CONSTBUFFER_ARRAY_HANDLE filtered_array = constbuffer_array_remove_empty_buffers(original_array);
+    
+    ///assert
+    ASSERT_IS_NOT_NULL(filtered_array);
+    
+    // Verify the filtered array contains only non-empty buffers
+    uint32_t filtered_count;
+    int result = constbuffer_array_get_buffer_count(filtered_array, &filtered_count);
+    ASSERT_ARE_EQUAL(int, 0, result);
+    ASSERT_ARE_EQUAL(uint32_t, 3, filtered_count);
```
Comment (dcristoloveanu) [inline tests/constbuffer_array_int/constbuffer_array_int.c:86]: No need for the intermediate result variable, this can be done in one line

File: tests/constbuffer_array_int/constbuffer_array_int.c [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+    ASSERT_IS_NOT_NULL(empty_buffer3);
+    
+    // Create array with mixed empty and non-empty buffers
+    CONSTBUFFER_HANDLE buffers[6];
+    buffers[0] = header_buffer;
+    buffers[1] = empty_buffer1;
+    buffers[2] = payload_buffer;
+    buffers[3] = empty_buffer2;
+    buffers[4] = footer_buffer;
+    buffers[5] = empty_buffer3;
```
[the remark this answers is not in the document]
Comment (dcristoloveanu) [reply tests/constbuffer_array_int/constbuffer_array_int.c:71]: Same comment for the other tests in the file.

File: tests/constbuffer_array_int/constbuffer_array_int.c (hunk 2 above)
Comment (dcristoloveanu) [reply tests/constbuffer_array_int/constbuffer_array_int.c:86]: Same comment for the other tests in the file

File: tests/constbuffer_array_int/constbuffer_array_int.c [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    CONSTBUFFER_DecRef(buffer4);
+    CONSTBUFFER_DecRef(empty4);
+    constbuffer_array_dec_ref(original_array);
+    constbuffer_array_dec_ref(filtered_array);
+}
+
+TEST_FUNCTION(constbuffer_array_remove_empty_buffers_with_only_empty_buffers_returns_empty_array)
+{
+    ///arrange
+    const uint32_t buffer_count = 5;
```
Comment (dcristoloveanu) [inline tests/constbuffer_array_int/constbuffer_array_int.c:273]: >5 [](http://example.com/codeflow?start=34&length=1)

No need to have 5, 2 is enough

File: devdoc/constbuffer_array_requirements.md [hunk 5]
```diff
@@ -19,6 +19,9 @@ MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_from_buff
 MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_empty);
 MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_from_array_array, const CONSTBUFFER_ARRAY_HANDLE*, buffer_arrays, uint32_t, buffer_array_count);
 
+/*transformation*/
```
Comment (dcristoloveanu) [inline devdoc/constbuffer_array_requirements.md:22]: This comment can be removed, adds no value.

File: inc/c_util/constbuffer_array.h [hunk 6]
```diff
@@ -28,6 +28,9 @@ MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_from_buff
 MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_empty);
 MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_from_array_array, const CONSTBUFFER_ARRAY_HANDLE*, buffer_arrays, uint32_t, buffer_array_count);
 
+/*transformation*/
```
Comment (dcristoloveanu) [inline inc/c_util/constbuffer_array.h:31]: Same here, likely comment can be removed.

File: src/constbuffer_array.c [hunk 7]
```diff
@@ -857,6 +857,94 @@ IMPLEMENT_MOCKABLE_FUNCTION(, const CONSTBUFFER_HANDLE*, constbuffer_array_get_c
...
+    {
+        /*Codes_SRS_CONSTBUFFER_ARRAY_88_001: [ If constbuffer_array_handle is NULL then constbuffer_array_remove_empty_buffers shall fail and return NULL. ]*/
+        LogError("Invalid arguments: CONSTBUFFER_ARRAY_HANDLE constbuffer_array_handle=%p", constbuffer_array_handle);
+        result = NULL;
+    }
+    else
+    {
+        /*Codes_SRS_CONSTBUFFER_ARRAY_88_002: [ constbuffer_array_remove_empty_buffers shall get the buffer count from constbuffer_array_handle. ]*/
+        if (constbuffer_array_handle->nBuffers == 0)
+        {
```
Comment (dcristoloveanu) [inline src/constbuffer_array.c:874]: I think we should not treat differently the case when the input const buffer array has 0 buffers. If we have 0 buffers we would figure out after examining the input const buffer array that the number of non empty buffers has not changed and we should simply return the same input const buffer array with its reference count incremented.

File: devdoc/constbuffer_array_requirements.md [hunk 8]
```diff
@@ -167,6 +169,34 @@ MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_from_arra
 
 **SRS_CONSTBUFFER_ARRAY_42_008: [** If there are any failures then `constbuffer_array_create_from_array_array` shall fail and return `NULL`. **]**
 
+### constbuffer_array_remove_empty_buffers
```
Comment (jebrando) [inline devdoc/constbuffer_array_requirements.md:172]: >constbuffer_array_remove_empty_buffers [](http://example.com/codeflow?start=4&length=38)

Should we have a function to tell us if there is an empty array in a buffer, since this function will always create a new array, it would be nice to know if we need to call this or not.

**Review by jebrando — APPROVED** [review]
:shipit:
